doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -0,0 +1,149 @@
+
+A *cost model* tries to predict a program’s behavior given information about its inputs. Given a program and some contextual information, a cost model produces an *estimated cost* for that query. Optimizers use cost models to make better informed decisions.
+
+Our cost model should be in terms of the dimensions that are relevant to us:
+
+- Latency (rehydration time; latency per update)
+- Storage (memory footprint/implicit storage)
+- Scalability (to what degree can this query be spread across workers)
+
+Our cost model should be *executable* and *accurate:*
+
+- Our cost model will be code that runs on actual MIR, producing a symbolic expression that represents the cost of the query along various dimensions. A user can use `EXPLAIN WITH(cost)`  to predict costs.
+- Our cost model should be tested on a variety of loads to ensure that our estimates aren’t total nonsense.
+
+# Motivation
+[motivation]: #motivation
+
+## Goals
+
+- Reify what we learn from various benchmarks/experiments.
+    - https://github.com/MaterializeInc/materialize/issues/17883
+    - https://github.com/MaterializeInc/materialize/issues/16511
+    - https://github.com/MaterializeInc/materialize/issues/15244
+    - https://github.com/MaterializeInc/materialize/issues/13140
+    - https://github.com/MaterializeInc/materialize/issues/12069
+    - https://github.com/MaterializeInc/materialize/issues/9157
+- Offer visibility/predicability.
+    - https://github.com/MaterializeInc/materialize/issues/14663
+    - https://github.com/MaterializeInc/materialize/issues/13046
+- Open opportunities for principled/well-informed choices in the optimizer.
+
+## Non-goals
+
+- Identify all opportunities to use the cost model in the optimizer.
+- Refactor the optimizer to be focused on the cost model.
+- Implement dynamic re-optimization based on changes in input statistics.
+- Collect all relevant statistics in storage and make them available to the cost model.
+
+# Explanation
+[explanation]: #explanation
+
+## Proposal
+
+A cost model is an abstract interpretation of the query: rather than running the query on relations and getting new relation out, we run the query on a symbolic description of its input relations and get a symbolic description of the result out.
+
+The general approach can be the same as `mz_transform::Typecheck`: recursively walk the `MirRelationExpr` with some contextual information, returning a symbolic term characterizing the query.
+
+### Context information
+
+Traditional databases track cardinality data and statistics about various columns. We will be interested in other statistics, too. Some possible examples:
+
+- How frequently does data arrive?
+- Is the given input physically monotonic or append-only? (I.e., there are no retractions.)
+- At what rate do we expect new data to arrive?
+- Some kind of locality property?
```

Comment:
I would say we should include key uniqueness and foreign key constraints here if possible.